### PR TITLE
Add an animated value for SvgLengthOrPercentageOrNumber.

### PR DESCRIPTION
**Pull request by mantaroh**
This is a PR for https://bugzilla.mozilla.org/show_bug.cgi?id=1390367

This patch will add an animated value for SvgLengthOrPercentageOrNumber in order to support SMIL animation with fill='freeze'.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors

There are already these tests in dom/smil/tests of gecko, this PR will enable these tests.
For detail, see https://bugzilla.mozilla.org/show_bug.cgi?id=1390367.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17288)

### CODE REVIEW

File: components/style/values/generics/svg.rs [hunk 1]
```diff
@@ -101,80 +102,68 @@ impl<ColorType: Parse, UrlPaintServer: Parse> Parse for SVGPaint<ColorType, UrlP
...
-impl<L, N> ComputeSquaredDistance for SvgLengthOrPercentageOrNumber<L, N>
+/// Following From implements use for converting animated value.
+impl<L, N> ToAnimatedValue for SvgLengthOrPercentageOrNumber<L, N>
     where
-        L: ComputeSquaredDistance + Copy + Into<NumberOrPercentage>,
-        N: ComputeSquaredDistance + Copy + Into<NumberOrPercentage>
+        L: Into<SvgStrokeNumberOrPercentageOrCalc> + From<LengthOrPercentage>,
+        N: Into<SvgStrokeNumberOrPercentageOrCalc> + From<Number>
 {
+    type AnimatedValue = SvgStrokeNumberOrPercentageOrCalc;
```
Comment (nox) [inline components/style/values/generics/svg.rs]: No, something named "length" shouldn't have a more specific animated value with "stroke" in its name. What exactly are you doing with this patch?

File: components/style/values/generics/svg.rs [hunk 2]
```diff
@@ -193,16 +182,76 @@ pub enum SVGLength<LengthType> {
     ContextValue,
 }
 
+impl<L> ToAnimatedValue for SVGLength<L>
```
Comment (nox) [inline components/style/values/generics/svg.rs]: This should be derived.

File: components/style/values/animated/svg.rs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+//! Animated types for CSS values in SVG
+
+use app_units::Au;
+use values::animated::{Animate, Procedure, ToAnimatedZero};
+use values::computed::{NonNegativeNumber, Number, Percentage};
+use values::computed::length::{CalcLengthOrPercentage, LengthOrPercentage};
+use values::computed::length::NonNegativeLengthOrPercentage;
+
+
+/// | number | percentage | calc| for stroke-*
```
Comment (nox) [inline components/style/values/animated/svg.rs]: This comment is weird.

File: components/style/values/animated/svg.rs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+//! Animated types for CSS values in SVG
+
+use app_units::Au;
+use values::animated::{Animate, Procedure, ToAnimatedZero};
+use values::computed::{NonNegativeNumber, Number, Percentage};
+use values::computed::length::{CalcLengthOrPercentage, LengthOrPercentage};
+use values::computed::length::NonNegativeLengthOrPercentage;
+
+
```
Comment (nox) [inline components/style/values/animated/svg.rs]: Nit: two many newlines.

File: components/style/values/animated/svg.rs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+                Ok(SvgStrokeNumberOrPercentageOrCalc::Percentage(this.animate(other, procedure)?))
+            },
+            // TODO: We need to support calc value.
+            // https://bugzilla.mozilla.org/show_bug.cgi?id=1386967
+            _ => Err(()),
+        }
+    }
+}
+
+impl ToAnimatedZero for SvgStrokeNumberOrPercentageOrCalc {
```
Comment (nox) [inline components/style/values/animated/svg.rs]: Derive this.

File: components/style/values/generics/svg.rs [hunk 6]
```diff
@@ -193,16 +182,76 @@ pub enum SVGLength<LengthType> {
...
-#[derive(Clone, ComputeSquaredDistance, Debug, PartialEq, HasViewportPercentage, ToAnimatedValue, ToComputedValue)]
+#[derive(Clone, ComputeSquaredDistance, Debug, PartialEq, HasViewportPercentage,  ToComputedValue)]
 pub enum SVGStrokeDashArray<LengthType> {
     /// `[ <length> | <percentage> | <number> ]#`
     Values(Vec<LengthType>),
     /// `context-value`
     ContextValue,
 }
 
+impl<L> ToAnimatedValue for SVGStrokeDashArray<L>
```
Comment (nox) [inline components/style/values/generics/svg.rs]: Derive this.

**Review by nox — CHANGES_REQUESTED** [review]

File: components/style/values/generics/svg.rs (hunk 1 above)
Comment (mantaroh) [reply components/style/values/generics/svg.rs]: Thanks for reviewing.
I addressed the patch:
 * Rename SvgStrokeNumberOrPercentageOrCalc to SvgNumberOrPercentageOrCalc.
 * Use ToAnimatedValue derive.

File: components/style/values/animated/svg.rs [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+use values::animated::{Animate, Procedure};
+use values::computed::{NonNegativeNumber, Number, Percentage};
+use values::computed::length::{CalcLengthOrPercentage, LengthOrPercentage};
+use values::computed::length::NonNegativeLengthOrPercentage;
+
+/// Stroke-* value support unit less value, so servo interpolate length value as number.
+#[derive(Clone, ComputeSquaredDistance, Copy, Debug, PartialEq, ToAnimatedZero)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+#[allow(missing_docs)]
+pub enum SvgNumberOrPercentageOrCalc {
```
Comment (nox) [inline components/style/values/animated/svg.rs]: I still don't get it, why isn't that named `SvgLengthOrPercentageOrNumber` just as the non-animated version?

File: components/style/values/animated/svg.rs [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+use app_units::Au;
+use values::animated::{Animate, Procedure};
+use values::computed::{NonNegativeNumber, Number, Percentage};
+use values::computed::length::{CalcLengthOrPercentage, LengthOrPercentage};
+use values::computed::length::NonNegativeLengthOrPercentage;
+
+/// Stroke-* value support unit less value, so servo interpolate length value as number.
+#[derive(Clone, ComputeSquaredDistance, Copy, Debug, PartialEq, ToAnimatedZero)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+#[allow(missing_docs)]
```
Comment (nox) [inline components/style/values/animated/svg.rs]: No.

File: components/style/values/animated/svg.rs [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+//! Animated types for CSS values in SVG
+
+use app_units::Au;
+use values::animated::{Animate, Procedure};
+use values::computed::{NonNegativeNumber, Number, Percentage};
+use values::computed::length::{CalcLengthOrPercentage, LengthOrPercentage};
+use values::computed::length::NonNegativeLengthOrPercentage;
+
+/// Stroke-* value support unit less value, so servo interpolate length value as number.
```
Comment (nox) [inline components/style/values/animated/svg.rs]: Why is this about `stroke-*` if the type name is about arbitrary numbers and percentages? The naming in this PR is still not good. Where exactly is this type used?

File: components/style/values/animated/svg.rs [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+            },
+            LengthOrPercentage::Calc(calc) => {
+                SvgNumberOrPercentageOrCalc::Calc(calc.into())
+            },
+        }
+    }
+
+}
+
+impl From<Number> for SvgNumberOrPercentageOrCalc {
```
Comment (nox) [inline components/style/values/animated/svg.rs]: Where is this conversion needed?

File: components/style/values/animated/svg.rs [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    fn from(num: Number) -> SvgNumberOrPercentageOrCalc {
+        SvgNumberOrPercentageOrCalc::Number(num)
+    }
+}
+
+impl From<SvgNumberOrPercentageOrCalc> for CalcLengthOrPercentage {
```
Comment (nox) [inline components/style/values/animated/svg.rs]: Where is this conversion needed?

File: components/style/values/animated/svg.rs [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+                CalcLengthOrPercentage::new(Au::from_f32_px(num), None),
+            SvgNumberOrPercentageOrCalc::Percentage(p) =>
+                CalcLengthOrPercentage::new(Au(0), Some(p)),
+            SvgNumberOrPercentageOrCalc::Calc(calc) =>
+                calc.into(),
+        }
+    }
+}
+
+impl Animate for SvgNumberOrPercentageOrCalc {
```
Comment (nox) [inline components/style/values/animated/svg.rs]: Derive this, you can put `#[animation(error)]` on the `Calc` variant.

File: components/style/values/generics/svg.rs [hunk 13]
```diff
@@ -101,80 +102,68 @@ impl<ColorType: Parse, UrlPaintServer: Parse> Parse for SVGPaint<ColorType, UrlP
...
-impl<L, N> ComputeSquaredDistance for SvgLengthOrPercentageOrNumber<L, N>
+/// Following From implements use for converting animated value.
+impl<L, N> ToAnimatedValue for SvgLengthOrPercentageOrNumber<L, N>
     where
-        L: ComputeSquaredDistance + Copy + Into<NumberOrPercentage>,
-        N: ComputeSquaredDistance + Copy + Into<NumberOrPercentage>
+        L: Into<SvgNumberOrPercentageOrCalc> + From<LengthOrPercentage>,
+        N: Into<SvgNumberOrPercentageOrCalc> + From<Number>
 {
+    type AnimatedValue = SvgNumberOrPercentageOrCalc;
```
Comment (nox) [inline components/style/values/generics/svg.rs]: Why does the name change?

File: components/style/values/generics/svg.rs [hunk 14]
```diff
@@ -101,80 +102,68 @@ impl<ColorType: Parse, UrlPaintServer: Parse> Parse for SVGPaint<ColorType, UrlP
...
-            },
-            (
-                &SvgLengthOrPercentageOrNumber::LengthOrPercentage(from),
-                &SvgLengthOrPercentageOrNumber::Number(to)
-            ) => {
-                from.into().compute_squared_distance(&to.into())
+    fn to_animated_value(self) -> Self::AnimatedValue {
+        match self {
+            SvgLengthOrPercentageOrNumber::LengthOrPercentage(lop) =>
+            {
```
Comment (nox) [inline components/style/values/generics/svg.rs]: No braces on their own line.

**Review by nox — CHANGES_REQUESTED** [review]

File: components/style/values/animated/svg.rs (hunk 9 above)
Comment (mantaroh) [reply components/style/values/animated/svg.rs]: Thank you so much!
I changed to these name to same to computed value.

File: components/style/values/animated/svg.rs (hunk 10 above)
Comment (mantaroh) [reply components/style/values/animated/svg.rs]: These conversions are for `ToAnimated` of `SvgLengthOrPercentageOrNumber`. So I moved these conversions to `generics/svg.rs`, and commented.

**Comment by nox** [comment]
The more I look at this patch, the less I understand what it is supposed to do. Why can't you just use `SvgLengthOrPercentageOrNumber` during animations?

**Comment by mantaroh** [comment]
@nox
Thank you for confirming this.

We can use `SvgLengthOrPercentageOrNumber` during animation, but SMIL with fill='freeze' uses finished animation value to specified value(i.e. if we specify <animate .. to='20px'>, finished animation value use this '20px'). In gecko, this animated property(stroke-*) has CSS_PROPERTY_NUMBERS_ARE_PIXELS flag, So finished animation value will unit-less number value[ [StyleAnimationValue.cpp](http://searchfox.org/mozilla-central/rev/51b3d67a5ec1758bd2fe7d7b6e75ad6b6b5da223/layout/style/StyleAnimationValue.cpp#4711) ].
But Servo will return this value with unit length for now since servo parse to LengthOrPercentage. So I think that we need to store animated value in order to keep unit-less value of animation result.

**Comment by birtles** [comment]
I haven't looked at these patches so I'm not sure how much this helps, but essentially the issue we have is that for `SvgLengthOrPercentageOrNumber` we call `convert_to_number_or_percentage` during _interpolation_ (i.e. `add_weighted`). For animation values that don't get interpolated (because we determined we don't need to), that conversion from lengths to numbers never happens and we serialize with 'px'.

In Gecko we don't have that problem because we convert lengths to numbers when we _create_ animation values, not when we interpolate them.

**Comment by mantaroh** [comment]
Oops, Sorry.
This PR contains the mis-merge code.(Especially, an use declaration.)
I'll update soon.

**Comment by mantaroh** [comment]
Thanks @birtles 
If animated value type is different from specified value type like stroke-*,  I think that servo need to store the converted value into animated value.

**Comment by nox** [comment]
> In Gecko we don't have that problem because we convert lengths to numbers when we create animation values, not when we interpolate them.

So in Gecko it never keeps the px unit? Even in specified and computed values?

**Comment by mantaroh** [comment]
No, Gecko keeps the px unit if value in specified and computed values, but animation value is always unit-less number.

When creating animation value, gecko converts animation value without the px unit if target property has  `CSS_PROPERTY_NUMBERS_ARE_PIXELS` flags. This flags added when defining properties, stroke-* has this flags.

**Comment by nox** [comment]
Is this patch obsolete?

**Comment by nox** [comment]
@mantaroh Ping?

**Comment by mantaroh** [comment]
@nox 
Sorry, Yes. This is already obsoleted patches.
Thanks.

**Pull request closed without merging**